

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Quickstart &mdash; mutyper  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/style.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Citing mutyper" href="../cite.html" />
    <link rel="prev" title="Installation" href="../install.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">User Guide</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Python-API-demo">Python API demo</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Command-line-interface-demo">Command line interface demo</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#variants-subcommand"><code class="docutils literal notranslate"><span class="pre">variants</span></code> subcommand</a></li>
<li class="toctree-l3"><a class="reference internal" href="#targets-subcommand"><code class="docutils literal notranslate"><span class="pre">targets</span></code> subcommand</a></li>
<li class="toctree-l3"><a class="reference internal" href="#spectra-subcommand"><code class="docutils literal notranslate"><span class="pre">spectra</span></code> subcommand</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ksfs-subcommand"><code class="docutils literal notranslate"><span class="pre">ksfs</span></code> subcommand</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../cite.html">Citing <code class="docutils literal notranslate"><span class="pre">mutyper</span></code></a></li>
</ul>
<p class="caption"><span class="caption-text">CLI Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../cmd.html">Command Line Interface</a></li>
</ul>
<p class="caption"><span class="caption-text">API Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../mutyper.html">Python API: <code class="docutils literal notranslate"><span class="pre">mutyper</span></code> package</a></li>
</ul>
<p class="caption"><span class="caption-text">Developer Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer.html">Open source code repository</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer.html#update-documentation">Update documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer.html#todo-list">Todo list</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">mutyper</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Quickstart</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/notebooks/quickstart.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Quickstart">
<h1>Quickstart<a class="headerlink" href="#Quickstart" title="Permalink to this headline">¬∂</a></h1>
<p>This notebook demonstrates how to use <code class="docutils literal notranslate"><span class="pre">mutyper</span></code> for computing mutation type data for VCF/BCF data</p>
<div class="section" id="Python-API-demo">
<h2>Python API demo<a class="headerlink" href="#Python-API-demo" title="Permalink to this headline">¬∂</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">mutyper</span>
</pre></div>
</div>
</div>
<p>Path to an ancestral FASTA file for human chromosome 1</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fasta</span> <span class="o">=</span> <span class="s1">&#39;../example_data/ancestor.fa&#39;</span>
</pre></div>
</div>
</div>
<p>Show the first 3 lines with a bash cell</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>head -3 <span class="o">{</span>fasta<span class="o">}</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&gt;ANCESTOR_for_chromosome:GRCh37:1:1:249250621:1
TTTGTGGGAgACTATTCCTCCCATCTGCAACAGCTGCCCCTGCTGACTGCCCTTCTCTCCTCCCTCTCATCCCAGAGAAACAgGTCAGCTGGGAGCTTCT
GCCCCCACTGCCTAGGGACCAACAGGGGCAGGAGGCAGTCACTGACCCCGAGACGTTTGCATCCTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGT
</pre></div></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">mutyper</span></code> package has a single class <code class="docutils literal notranslate"><span class="pre">Ancestor</span></code>. We instantiate an <code class="docutils literal notranslate"><span class="pre">Ancestor</span></code> object using our FASTA file - We‚Äôre interested in 3-mer context, which we can indicate with the <code class="docutils literal notranslate"><span class="pre">k</span></code> keyword argument (by default it will be 3 though). - The chromosome name is the 3rd field delimited by <code class="docutils literal notranslate"><span class="pre">:</span></code> in the FASTA record id. We use the <code class="docutils literal notranslate"><span class="pre">key_function</span></code> keyword argument to deal with this.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ancestor</span> <span class="o">=</span> <span class="n">mutyper</span><span class="o">.</span><span class="n">Ancestor</span><span class="p">(</span><span class="n">fasta</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">key_function</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
</div>
<p>We can inspect FASTA record names, showing that the FASTA contains two records, chromosome <code class="docutils literal notranslate"><span class="pre">'1'</span></code> and <code class="docutils literal notranslate"><span class="pre">'2'</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ancestor</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
odict_keys([&#39;1&#39;, &#39;2&#39;])
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ancestor</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
FastaRecord(&#34;1&#34;)
</pre></div></div>
</div>
<p>and we can Pythonically slice sequences via fast random access without loading into memory (leaning on <code class="docutils literal notranslate"><span class="pre">`pyfaidx</span></code> &lt;<a class="reference external" href="https://pythonhosted.org/pyfaidx/">https://pythonhosted.org/pyfaidx/</a>&gt;`__ under the hood):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">start</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">end</span> <span class="o">=</span> <span class="mi">300</span>
<span class="n">ancestor</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">][</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&gt;1:101-300
GCCCCCACTGCCTAGGGACCAACAGGGGCAGGAGGCAGTCACTGACCCCGAGACGTTTGCATCCTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGTTGGTTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAGCCTCTTAAGAACACAGTGGCGCAGGCTGGGTGGAGCCGTCCCCCCATGgAG
</pre></div></div>
</div>
<p>We can also access these <code class="docutils literal notranslate"><span class="pre">FastaRecord</span></code> slices as string-like biopython <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ancestor</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">][</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;GCCCCCACTGCCTAGGGACCAACAGGGGCAGGAGGCAGTCACTGACCCCGAGACGTTTGCATCCTGCACAGCTAGAGATCCTTTATTAAAAGCACACTGTTGGTTTCTGCTCAGTTCTTTATTGATTGGTGTGCCGTTTTCTCTGGAAGCCTCTTAAGAACACAGTGGCGCAGGCTGGGTGGAGCCGTCCCCCCATGgAG&#39;
</pre></div></div>
</div>
<p><strong>The ``mutation_type`` method</strong> allows us to specify a SNP by the usual CHROM, POS, REF, ALT, and returns the correctly polarized mutation type as a tuple of ancestral kmer and derived kmer.</p>
<p>First, consider the triplet context at site 100 on chromosome 1:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ancestor</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">][</span><span class="mi">98</span><span class="p">:</span><span class="mi">101</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&gt;1:99-101
CTG
</pre></div></div>
</div>
<p>Suppose a biallelic SNP at this site segregates in a population with reference allele T and alternative allele A. The mutation type of this SNP is</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ancestor</span><span class="o">.</span><span class="n">mutation_type</span><span class="p">(</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="mi">99</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(&#39;CAG&#39;, &#39;CTG&#39;)
</pre></div></div>
</div>
<p>Note that, by default, mutation types are collapsed by reverse complementation such that the ancestral state at the target site is A or C. So in the above, the ancestral state T caused the ancestral triplet CTG to be reversed and complemented to CAG.</p>
<p>If the reference and alternative states are both distinct from the ancestral state, the site is not biallelic (an infinite sites violation), and the mutation type is returned as ambiguous.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ancestor</span><span class="o">.</span><span class="n">mutation_type</span><span class="p">(</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="mi">99</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(None, None)
</pre></div></div>
</div>
<p><strong>The ``region_contexts`` method</strong> returns a generator of ancestral contexts (by default collapsed as described above) over the positions in a region specified as in BED file format CHROM, START, END</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">start</span> <span class="o">=</span> <span class="mi">50</span>
<span class="n">end</span> <span class="o">=</span> <span class="mi">100</span>
<span class="k">for</span> <span class="n">context</span> <span class="ow">in</span> <span class="n">ancestor</span><span class="o">.</span><span class="n">region_contexts</span><span class="p">(</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">context</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CCC CCT AAG GAA TCT GAG TCT GAG TCC CCT GAG TCC CCC CCT GAG TCT GAG TCA CAT GAT TCC CCC CCA CAG TCT GAG TCT GAA AAA AAC ACA None None None GAC TCA CAG GCT GCT CAG CCA CCC TCC GAG GCT GCT AAG GAA TCT CAG
</pre></div></div>
</div>
<p>Note that FASTA sites that are not nucleotide characters have non-identified ancestral state, so context is <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<p>This generator method is used by the <strong>the ``targets`` method</strong> to compute the masked genomic target size for each <span class="math notranslate nohighlight">\(k\)</span>-mer context from a BED mask file. This may be useful for normalizing spectra in different genomic regions, or calibrating mutation rates.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">bed</span> <span class="o">=</span> <span class="s1">&#39;../example_data/mask.bed&#39;</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>cat <span class="o">{</span>bed<span class="o">}</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
1       25      50
1       150     175
2       50      100
</pre></div></div>
</div>
<p>Under the hood it loops over BED file entries and updates a <code class="docutils literal notranslate"><span class="pre">Counter</span></code> object for the different triplets</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ancestor</span><span class="o">.</span><span class="n">targets</span><span class="p">(</span><span class="n">bed</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Counter({&#39;CAG&#39;: 10,
         &#39;GCA&#39;: 9,
         &#39;CAA&#39;: 2,
         &#39;AAC&#39;: 3,
         &#39;ACA&#39;: 2,
         &#39;GCT&#39;: 7,
         &#39;GCC&#39;: 4,
         &#39;CCC&#39;: 3,
         &#39;CCT&#39;: 3,
         &#39;TCA&#39;: 2,
         &#39;GAC&#39;: 2,
         &#39;ACT&#39;: 1,
         &#39;GAG&#39;: 4,
         &#39;TCT&#39;: 7,
         &#39;ACG&#39;: 2,
         &#39;AAA&#39;: 7,
         &#39;CAT&#39;: 5,
         &#39;GAT&#39;: 5,
         &#39;TCC&#39;: 3,
         &#39;CAC&#39;: 1,
         &#39;TAG&#39;: 4,
         &#39;AAT&#39;: 3,
         &#39;TAA&#39;: 4,
         &#39;CCA&#39;: 4,
         &#39;GAA&#39;: 1,
         &#39;AAG&#39;: 1,
         &#39;ACC&#39;: 1})
</pre></div></div>
</div>
<p>We can also call it without the BED file mask to compute the unmasked target sizes of the ancestral FASTA</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ancestor</span><span class="o">.</span><span class="n">targets</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Counter({&#39;AAA&#39;: 6197,
         &#39;CAA&#39;: 3408,
         &#39;ACA&#39;: 3920,
         &#39;CAC&#39;: 3260,
         &#39;CCA&#39;: 3764,
         &#39;CCC&#39;: 3234,
         &#39;TCC&#39;: 3317,
         &#39;ACT&#39;: 2998,
         &#39;TAG&#39;: 2210,
         &#39;TAT&#39;: 3279,
         &#39;AAT&#39;: 4240,
         &#39;GAA&#39;: 3861,
         &#39;CCT&#39;: 3811,
         &#39;GAG&#39;: 3601,
         &#39;CAT&#39;: 3456,
         &#39;GAT&#39;: 2370,
         &#39;TCT&#39;: 4319,
         &#39;CAG&#39;: 4423,
         &#39;GCA&#39;: 3182,
         &#39;AAC&#39;: 2688,
         &#39;GCT&#39;: 2972,
         &#39;GCC&#39;: 2694,
         &#39;TCA&#39;: 3730,
         &#39;GAC&#39;: 1998,
         &#39;AAG&#39;: 3860,
         &#39;ACC&#39;: 2378,
         &#39;CCG&#39;: 717,
         &#39;TCG&#39;: 450,
         &#39;ACG&#39;: 554,
         &#39;TAA&#39;: 3527,
         &#39;GCG&#39;: 592,
         &#39;TAC&#39;: 1927})
</pre></div></div>
</div>
</div>
<div class="section" id="Command-line-interface-demo">
<h2>Command line interface demo<a class="headerlink" href="#Command-line-interface-demo" title="Permalink to this headline">¬∂</a></h2>
<p>Note that command line cells begin with <code class="docutils literal notranslate"><span class="pre">!</span></code> in Jupyter notebooks.</p>
<p>Display usage information for the mutyper command</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>mutyper -h
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
usage: mutyper [-h] {variants,targets,spectra,ksfs} ...

mutyper: ancestral kmer mutation types for variant data

optional arguments:
  -h, --help            show this help message and exit

subcommands:
  specify one of these

  {variants,targets,spectra,ksfs}
                        additional help available for each subcommand
</pre></div></div>
</div>
<div class="section" id="variants-subcommand">
<h3><code class="docutils literal notranslate"><span class="pre">variants</span></code> subcommand<a class="headerlink" href="#variants-subcommand" title="Permalink to this headline">¬∂</a></h3>
<p>Usage:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>mutyper variants -h
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
usage: mutyper variants [-h] [--k K] [--target TARGET] [--sep SEP]
                        [--chrom_pos CHROM_POS] [--strand_file STRAND_FILE]
                        [--strict]
                        fasta vcf

adds mutation_type to VCF/BCF INFO, polarizes REF/ALT/AC according to
ancestral/derived states, and stream to stdout

positional arguments:
  fasta                 path to ancestral FASTA
  vcf                   VCF/BCF file created by the variants subcommand (&#34;-&#34;
                        for stdin)

optional arguments:
  -h, --help            show this help message and exit
  --k K                 k-mer context size (default 3)
  --target TARGET       0-based mutation target position in kmer (default
                        middle)
  --sep SEP             field delimiter in FASTA headers (default &#34;:&#34;)
  --chrom_pos CHROM_POS
                        0-based chromosome field position in FASTA headers
                        (default 2)
  --strand_file STRAND_FILE
                        path to bed file with regions where reverse strand
                        defines mutation context, e.g. direction of
                        replication or transcription (default collapse reverse
                        complements)
  --strict              only uppercase nucleotides in FASTA considered
                        ancestrally identified
</pre></div></div>
</div>
<p>Path to a truncated VCF file for chromosome 1 from the 1000 Genomes Project.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vcf</span> <span class="o">=</span> <span class="s1">&#39;../example_data/snps.vcf&#39;</span>
</pre></div>
</div>
</div>
<p>We‚Äôll use <code class="docutils literal notranslate"><span class="pre">`bcftools</span> <span class="pre">view</span></code> &lt;<a class="reference external" href="http://samtools.github.io/bcftools/bcftools.html#view">http://samtools.github.io/bcftools/bcftools.html#view</a>&gt;`__ to display the first few variants, omitting header and genotype information.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>bcftools view -HG <span class="o">{</span>vcf<span class="o">}</span> <span class="p">|</span> head -5
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
1       10177   .       A       AC      100     PASS    AC=2130;AF=0.425319;AN=5008;NS=2504;DP=103152;EAS_AF=0.3363;AMR_AF=0.3602;AFR_AF=0.4909;EUR_AF=0.4056;SAS_AF=0.4949;AA=|||unknown(NO_COVERAGE)
1       10235   .       T       TA      100     PASS    AC=6;AF=0.00119808;AN=5008;NS=2504;DP=78015;EAS_AF=0;AMR_AF=0.0014;AFR_AF=0;EUR_AF=0;SAS_AF=0.0051;AA=|||unknown(NO_COVERAGE)
1       10352   rs145072688     T       TA      100     PASS    AC=2191;AF=0.4375;AN=5008;NS=2504;DP=88915;EAS_AF=0.4306;AMR_AF=0.4107;AFR_AF=0.4788;EUR_AF=0.4264;SAS_AF=0.4192;AA=|||unknown(NO_COVERAGE)
1       10505   .       A       T       100     PASS    AC=1;AF=0.000199681;AN=5008;NS=2504;DP=9632;EAS_AF=0;AMR_AF=0;AFR_AF=0.0008;EUR_AF=0;SAS_AF=0;AA=.|||
1       10506   .       C       G       100     PASS    AC=1;AF=0.000199681;AN=5008;NS=2504;DP=9676;EAS_AF=0;AMR_AF=0;AFR_AF=0.0008;EUR_AF=0;SAS_AF=0;AA=.|||
[main_vcfview] Error: cannot write to (null)
</pre></div></div>
</div>
<p>Now we use the <code class="docutils literal notranslate"><span class="pre">variants</span></code> subcommand, and pipe to <code class="docutils literal notranslate"><span class="pre">bcftools</span></code> for display. Notice there is an additional INFO field <code class="docutils literal notranslate"><span class="pre">mutation_type</span></code>, and only biallelic SNPs are output. Also note that REF/ALT state and INFO/AC fields are polarized according to ancestral/derived allele.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>mutyper variants <span class="o">{</span>fasta<span class="o">}</span> <span class="o">{</span>vcf<span class="o">}</span> <span class="p">|</span> bcftools view -HG <span class="p">|</span> head -5
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
1       10505   .       A       T       100     PASS    AC=5007;AF=0.000199681;AN=5008;NS=2504;DP=9632;EAS_AF=0;AMR_AF=0;AFR_AF=0.0008;EUR_AF=0;SAS_AF=0;AA=.|||;mutation_type=GAG&gt;GTG
1       10506   .       C       G       100     PASS    AC=1;AF=0.000199681;AN=5008;NS=2504;DP=9676;EAS_AF=0;AMR_AF=0;AFR_AF=0.0008;EUR_AF=0;SAS_AF=0;AA=.|||;mutation_type=TCT&gt;TGT
1       10542   .       A       G       100     PASS    AC=5007;AF=0.000199681;AN=5008;NS=2504;DP=9007;EAS_AF=0.001;AMR_AF=0;AFR_AF=0;EUR_AF=0;SAS_AF=0;AA=.|||;mutation_type=CAG&gt;CGG
1       10642   .       A       G       100     PASS    AC=4987;AF=0.00419329;AN=5008;NS=2504;DP=1360;EAS_AF=0.003;AMR_AF=0.0014;AFR_AF=0.0129;EUR_AF=0;SAS_AF=0;AA=.|||;mutation_type=CAT&gt;CGT
1       11008   .       C       G       100     PASS    AC=441;AF=0.0880591;AN=5008;NS=2504;DP=2232;EAS_AF=0.0367;AMR_AF=0.0965;AFR_AF=0.1346;EUR_AF=0.0885;SAS_AF=0.0716;AA=.|||;mutation_type=CCT&gt;CGT
</pre></div></div>
</div>
</div>
<div class="section" id="targets-subcommand">
<h3><code class="docutils literal notranslate"><span class="pre">targets</span></code> subcommand<a class="headerlink" href="#targets-subcommand" title="Permalink to this headline">¬∂</a></h3>
<p>Usage:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>mutyper targets -h
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
usage: mutyper targets [-h] [--k K] [--target TARGET] [--sep SEP]
                       [--chrom_pos CHROM_POS] [--strand_file STRAND_FILE]
                       [--strict] [--bed BED]
                       fasta

compute ùëò-mer target sizes and stream to stdout

positional arguments:
  fasta                 path to ancestral FASTA

optional arguments:
  -h, --help            show this help message and exit
  --k K                 k-mer context size (default 3)
  --target TARGET       0-based mutation target position in kmer (default
                        middle)
  --sep SEP             field delimiter in FASTA headers (default &#34;:&#34;)
  --chrom_pos CHROM_POS
                        0-based chromosome field position in FASTA headers
                        (default 2)
  --strand_file STRAND_FILE
                        path to bed file with regions where reverse strand
                        defines mutation context, e.g. direction of
                        replication or transcription (default collapse reverse
                        complements)
  --strict              only uppercase nucleotides in FASTA considered
                        ancestrally identified
  --bed BED             path to BED file mask (&#34;-&#34; for stdin)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>mutyper targets <span class="o">{</span>fasta<span class="o">}</span> <span class="p">|</span> head
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AAA     6331
AAC     2775
AAG     3947
AAT     4335
ACA     4066
ACC     2440
ACG     632
ACT     3082
CAA     3506
CAC     3391
</pre></div></div>
</div>
<p>The optional argument <code class="docutils literal notranslate"><span class="pre">--bed</span></code> resticts target size computation based on a BED mask file, and is useful of normalizing mutation spectra according to genomic region, or calibrating mutaiton rates.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>mutyper targets <span class="o">{</span>fasta<span class="o">}</span> --bed <span class="o">{</span>bed<span class="o">}</span> <span class="p">|</span> head
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AAA     7
AAC     3
AAG     1
AAT     3
ACA     2
ACC     1
ACG     2
ACT     1
CAA     2
CAC     1
</pre></div></div>
</div>
</div>
<div class="section" id="spectra-subcommand">
<h3><code class="docutils literal notranslate"><span class="pre">spectra</span></code> subcommand<a class="headerlink" href="#spectra-subcommand" title="Permalink to this headline">¬∂</a></h3>
<p>Usage:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>mutyper spectra -h
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
usage: mutyper spectra [-h] vcf

compute mutation spectra for each sample in VCF/BCF with mutation_type data
and stream to stdout

positional arguments:
  vcf         VCF/BCF file created by the variants subcommand (&#34;-&#34; for stdin)

optional arguments:
  -h, --help  show this help message and exit
</pre></div></div>
</div>
<p>We first add mutation type information with the <code class="docutils literal notranslate"><span class="pre">variants</span></code> subcommand, then pipe into the <code class="docutils literal notranslate"><span class="pre">spectra</span></code> subcommand to generate mutation spectra for all samples, and finally crop the output with <code class="docutils literal notranslate"><span class="pre">head</span></code> and <code class="docutils literal notranslate"><span class="pre">cut</span></code> for display.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>mutyper variants <span class="o">{</span>fasta<span class="o">}</span> <span class="o">{</span>vcf<span class="o">}</span> <span class="p">|</span> mutyper spectra - <span class="p">|</span> head <span class="p">|</span> cut -f-5
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
sample  AAA&gt;AGA      AAT&gt;AGT      ACA&gt;AAA      ACA&gt;AGA
HG00096 0       0       0       2
HG00097 0       0       0       2
HG00099 0       0       0       2
HG00100 0       0       0       2
HG00101 0       0       0       2
HG00102 0       0       0       2
HG00103 0       0       0       2
HG00105 0       0       0       2
HG00106 0       0       0       2
</pre></div></div>
</div>
</div>
<div class="section" id="ksfs-subcommand">
<h3><code class="docutils literal notranslate"><span class="pre">ksfs</span></code> subcommand<a class="headerlink" href="#ksfs-subcommand" title="Permalink to this headline">¬∂</a></h3>
<p>Usage:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>mutyper spectra -h
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
usage: mutyper spectra [-h] vcf

compute mutation spectra for each sample in VCF/BCF with mutation_type data
and stream to stdout

positional arguments:
  vcf         VCF/BCF file created by the variants subcommand (&#34;-&#34; for stdin)

optional arguments:
  -h, --help  show this help message and exit
</pre></div></div>
</div>
<p>Similar to the previous subcommand, but now we are generating a sample frequency spectrum (SFS) for each mutation type.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>mutyper variants <span class="o">{</span>fasta<span class="o">}</span> <span class="o">{</span>vcf<span class="o">}</span> <span class="p">|</span> mutyper ksfs - <span class="p">|</span> head <span class="p">|</span> cut -f-5
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
sample_frequency        AAA&gt;AGA      AAT&gt;AGT      ACA&gt;AAA      ACA&gt;AGA
1       0       0       0       0
2       0       0       0       0
3       0       0       0       0
4       0       0       0       0
5       0       0       0       0
6       0       0       0       0
7       0       0       0       0
8       0       0       0       0
9       0       0       0       0
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../cite.html" class="btn btn-neutral float-right" title="Citing mutyper" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../install.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, William DeWitt

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>